### PR TITLE
CC-1318 Add support for the --previous flag for the test command

### PR DESCRIPTION
**Pull request by libmartinito**

### CODE REVIEW

File: cmd/codecrafters/main.go [hunk 1]
```diff
@@ -23,11 +23,16 @@ USAGE
   $ codecrafters [command]
 
 EXAMPLES
-  $ codecrafters test     # Run tests without committing changes
-  $ codecrafters submit   # Commit changes & submit to move to next step
+  $ codecrafters test              # Run tests without committing changes
+  $ codecrafters test --previous   # Run tests for the current stage and all previous stages in ascending order
```
Comment (rohitpaulk) [inline cmd/codecrafters/main.go:27]:
```suggestion
  $ codecrafters test --previous   # Run tests for previous stages without committing changes
```

Let's just say this for now - in the future we'll change the actual behaviour to match this.

File: cmd/codecrafters/main.go [hunk 2]
```diff
@@ -23,11 +23,16 @@ USAGE
...
+  $ codecrafters submit            # Commit changes & submit to move to next step
 
 COMMANDS
   test:        Run tests without committing changes
+
+               Flags for test:
+               --previous   Run tests for the current stage and all previous stages in ascending order
```
Comment (rohitpaulk) [inline cmd/codecrafters/main.go:34]: Let's remove the `flags` section from here for now - will think about how to do proper flags per command. We need a better system for this - something like `codecrafters test --help`.

File: cmd/codecrafters/main.go [hunk 3]
```diff
@@ -76,7 +81,11 @@ func run() error {
 
 	switch cmd {
 	case "test":
-		return commands.TestCommand(ctx)
+		testCmd := flag.NewFlagSet("test", flag.ExitOnError)
+		previous := testCmd.Bool("previous", false, "run tests for the current stage and all previous stages in ascending order")
```
Comment (rohitpaulk) [inline cmd/codecrafters/main.go:85]:
```suggestion
		shouldTestPrevious := testCmd.Bool("previous", false, "run tests for the current stage and all previous stages in ascending order")
```

File: cmd/codecrafters/main.go [hunk 4]
```diff
@@ -76,7 +81,11 @@ func run() error {
...
+		testCmd.Parse(flag.Args()[1:]) // parse the args after the test command
```
Comment (rohitpaulk) [inline cmd/codecrafters/main.go:86]: Just curious - what happens if a user makes a typo here? If they do `codecrafters test --previos` for example? Will this complain? (Ideally it should)

File: internal/utils/codecrafters_client.go [hunk 5]
```diff
@@ -88,13 +88,20 @@ func (c CodecraftersClient) headers() map[string]string {
...
-func (c CodecraftersClient) CreateSubmission(repositoryId string, commitSha string) (CreateSubmissionResponse, error) {
+func (c CodecraftersClient) CreateSubmission(repositoryId string, commitSha string, previous bool) (CreateSubmissionResponse, error) {
+	requestBody := map[string]interface{}{
+		"repository_id":            repositoryId,
+		"commit_sha":               commitSha,
+		"should_auto_advance":      false,
+		"stage_selection_strategy": "current_and_previous_descending",
+	}
+
+	if previous {
```
Comment (rohitpaulk) [inline internal/utils/codecrafters_client.go:99]: This decision should be one layer above, in the caller. A "client" for core shouldn't be aware of anything other than what core supports. We should accept `stageSelectionStrategy` here, and have the caller decide what value to pass in.

**Review by rohitpaulk — CHANGES_REQUESTED** [review]
Added q

File: cmd/codecrafters/main.go (hunk 4 above)
Comment (libmartinito) [reply cmd/codecrafters/main.go:82]: Screenshot 2024-06-25 at 6 37 02 PM

tried it out and it does complain if a nonexistent flag is used for test

File: cmd/codecrafters/main.go [hunk 6]
```diff
@@ -23,8 +23,9 @@ USAGE
   $ codecrafters [command]
 
 EXAMPLES
-  $ codecrafters test     # Run tests without committing changes
-  $ codecrafters submit   # Commit changes & submit to move to next step
+  $ codecrafters test              # Run tests without committing changes
+  $ codecrafters test --previous   # Run tests for the current stage and all previous stages without comitting changes
```
Comment (rohitpaulk) [inline cmd/codecrafters/main.go:27]:
```suggestion
  $ codecrafters test --previous   # Run tests for the current stage and all previous stages without committing changes
```

**Review by rohitpaulk — APPROVED** [review]

**Pull request merged**